AmplifyPlugins/API/AWSAPICategoryPlugin/Operation/AWSGraphQLOperation.swift

```diff
@@ -100,35 +100,62 @@ final public class AWSGraphQLOperation<R: Decodable>: GraphQLOperation<R> {
         }
 
         // Create request
-        let urlRequest = GraphQLOperationRequestUtils.constructRequest(with: endpointConfig.baseURL,
-                                                                       requestPayload: requestPayload)
+        let urlRequest = GraphQLOperationRequestUtils.constructRequest(
+            with: endpointConfig.baseURL,
+            requestPayload: requestPayload)
 
         // Intercept request
-        let finalRequest = requestInterceptors.reduce(urlRequest) { (request, interceptor) -> URLRequest in
-            do {
-                return try interceptor.intercept(request)
-            } catch let error as APIError {
-                dispatch(result: .failure(error))
-                cancel()
-                return request
-            } catch {
-                dispatch(result: .failure(APIError.operationError("Failed to intercept request fully.",
-                                                                  "Something wrong with the interceptor",
-                                                                  error)))
-                cancel()
-                return request
+        chainInterceptors(iterator: requestInterceptors.makeIterator(),
+                          request: urlRequest) { finalRequest in
+            if self.isCancelled {
+                self.finish()
+                return
             }
+
+            // Begin network task
+            Amplify.API.log.debug("Starting network task for \(self.request.operationType) \(self.id)")
+            let task = self.session.dataTaskBehavior(with: finalRequest)
+            self.mapper.addPair(operation: self, task: task)
+            task.resume()
         }
+    }
+
+    private func chainInterceptors<I: IteratorProtocol>(
+        iterator: I,
+        request: URLRequest,
+        completion: @escaping (URLRequest) -> Void
+    ) where I.Element == URLRequestInterceptor {
 
         if isCancelled {
-            finish()
+            completion(request)
             return
         }
 
-        // Begin network task
-        Amplify.API.log.debug("Starting network task for \(request.operationType) \(id)")
-        let task = session.dataTaskBehavior(with: finalRequest)
-        mapper.addPair(operation: self, task: task)
-        task.resume()
+        var mutableIterator = iterator
+        guard let interceptor = mutableIterator.next() else {
+            completion(request)
+            return
+        }
+        interceptor.intercept(request) { result in
+            do {
+                let interceptedRequest = try result.get()
+                completion(interceptedRequest)
```

Comment:
should we call chainInterceptors(mutableIterator, interceptedRequest) here?

Reply:
oops! thank you for catching this.